runtime/queries/metricsview_aggregation.go

```diff
@@ -579,32 +572,88 @@ func (q *MetricsViewAggregation) buildMetricsAggregationSQL(mv *runtimev1.Metric
 			limitClause,                     // 8
 		)
 	} else {
-		sql = fmt.Sprintf("SELECT %s FROM %s %s %s %s %s %s %s OFFSET %d",
-			strings.Join(selectCols, ", "),
-			safeName(mv.Table),
-			strings.Join(unnestClauses, ""),
-			whereClause,
-			groupClause,
-			havingClause,
-			orderClause,
-			limitClause,
-			q.Offset,
-		)
-	}
+		/*
+			Example:
+			SELECT t.d1, t.d2, t.d3, t2.m1 (SELECT d1, d2, d3, m1 FROM t WHERE ...  GROUP BY d1, d2, d3 HAVING ... ) t LEFT JOIN (
+				SELECT d1, d2, d3, m1 FROM t WHERE ... AND (d4 = 'Safari') GROUP BY d1, d2, d3 HAVING ...
+			)  t2 ON (COALESCE(t.d1, 'val') = COALESCE(t2.d1, 'val') and COALESCE(t.d2, 'val') = COALESCE(t2.d2, 'val') and ...) ORDER BY ...
+
+			This JOIN mirrors functionality of SELECT d1, d2, d3, m1 FILTER (WHERE d4 = 'Safari') FROM t WHERE... GROUP BY d1, d2, d3
+			bacause FILTER cannot be applied for arbitrary measure, ie sum(a)/1000
+		*/
+		if len(q.Measures) == 1 && q.Measures[0].Filter != nil {
+			joinConditions := make([]string, 0, len(q.Dimensions))
+			selfJoinCols := make([]string, 0, len(q.Dimensions)+1)
+
+			selfJoinTableAlias := tempName("self_join")
+			nonNullValue := tempName("non_null")
+			for _, d := range q.Dimensions {
+				joinConditions = append(joinConditions, fmt.Sprintf("COALESCE(%[1]s.%[2]s, '%[4]s') = COALESCE(%[3]s.%[2]s, '%[4]s')", mv.Table, safeName(d.Name), selfJoinTableAlias, nonNullValue))
+				selfJoinCols = append(selfJoinCols, fmt.Sprintf("%s.%s", safeName(mv.Table), safeName(d.Name)))
+			}
+			selfJoinCols = append(selfJoinCols, fmt.Sprintf("%[1]s.%[2]s as %[3]s", selfJoinTableAlias, safeName(q.Measures[0].Name), safeName(q.Measures[0].Name)))
+			measureExpression, filterArgs, err := buildExpression(mv, q.Measures[0].Filter, nil, dialect)
+			if err != nil {
+				return "", nil, err
+			}
 
-	return sql, args, nil
-}
+			if whereClause == "" {
+				whereClause = "WHERE 1=1"
+			}
 
-func applyFilter(mv *runtimev1.MetricsViewSpec, expr string, filter *runtimev1.Expression, args *[]any, dialect drivers.Dialect) (string, error) {
-	if filter != nil {
-		exprClause, exprArgs, err := buildExpression(mv, filter, nil, dialect)
-		if err != nil {
-			return "", err
+			measureWhereClause := whereClause + fmt.Sprintf(" AND (%s)", measureExpression)
+			if len(sortingCriteria) > 0 {
+				jsc := make([]string, 0, len(sortingCriteria))
+				for _, v := range sortingCriteria {
+					jsc = append(jsc, fmt.Sprintf("%s.%s", safeName(mv.Table), v))
+				}
+				orderClause = "ORDER BY " + strings.Join(jsc, ", ")
+			}
+			sql = fmt.Sprintf(`
+					SELECT %[1]s FROM (
+						SELECT %[10]s FROM %[2]s %[3]s %[4]s %[5]s %[6]s 
+					) %[2]s 
+					LEFT JOIN (
+						SELECT %[10]s FROM %[2]s %[3]s %[9]s %[5]s %[6]s
+					) %[7]s 
+					ON (%[8]s) %[13]s %[11]s  OFFSET %[12]d
```

Comment:
I might be missing something, but the `having` clause seems to be applied in the sub-selects, but shouldn’t it be outside, after the join? For example, if you have having `measure1 < 10`, some dimension values might be excluded from the first sub-selects which are included in the second sub-select.